### PR TITLE
Register un-register git repos

### PR DESCRIPTION
**Pull request by Rohitrajak1807**
### This PR adds the ability to register a git repository with argocd for use with arlon and adds two sub-commands `git register` and `git unregister`

`git register` tries to replicate `argocd repo add`. When we run `arlon git register` it requires a repo URL, the username, the access token and an optional alias(which defaults to “default”)- if a “default” alias already exists, the repo isn’t registered with argocd and the alias creation fails saying that the default alias already exists otherwise, the repo is registered with argocd. Lastly we also write this repo info to  JSON file. This contains two pieces of information for each repo- it’s URL and the alias.
The structure of the file is as shown:
```json
{
    "default": {   // only to simplify the code. It avoids iterating through the repos array to find the default context
        "url": "",
        "alias": "default"
    },
    "repos": [
        {
            "url": "",
            "alias": "default"
        },
        {
            "url": "",
            "alias": "not_default"
       }, ...
    ]
}
```
When we run `arlon git unregister ALIAS`, it removes that entry from the local file. However, it does NOT remove the repository from argocd. When the “default” alias is deleted, we also clear the “default” entry from the JSON file.

Currently the config file is stored in the default `argocd` config file directory, under the filename `repoctx`.

This feature allows us to skip specifying git repository links and refer to them using an alias.

For example:
We could make the deploy command really short and simple.
From this: 
`bin/arlon cluster deploy --repo-url https://github.com/jayanth-tjvrr/nodelet-test --cluster-name nodelet-cluster-3 --profile static-1 --cluster-spec capi-nodelet-3`
To this: 
`bin/arlon cluster deploy --repo-alias prod --cluster-name nodelet-cluster-3 --profile static-1 --cluster-spec capi-nodelet-3`
Or even this: 
`bin/arlon cluster deploy --cluster-name nodelet-cluster-3 --profile static-1 --cluster-spec capi-nodelet-3` 
(when we don’t specify an alias, I would want it to pick the alias named “default”)

#### Breaking changes
- `profile create` subcommand now has a boolean flag `--static` which is used to indicate if a profile is static or not. It's set to `false` by default.

#### Testing
Manual:
```
➜  more_arlon git:(main) bin/arlon bundle create guestbook-dyn --tags applications --desc "guestbook app (dynamic)" --repo-path bundles/guestbook
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon bundle create guestbook-dyn1 --tags applications --desc "guestbook app (dynamic)" --repo-path bundles/guestbook --repo-alias rrajak
➜  more_arlon git:(main) bin/arlon bundle create guestbook-static --tags applications --desc "guestbook app" --from-file examples/bundles/guestbook.yaml
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon bundle list
NAME              TYPE     TAGS          REPO                                          PATH               REVISION  SRCTYPE  DESCRIPTION
guestbook-dyn     dynamic  applications  https://github.com/Rohitrajak1807/arln-oss    bundles/guestbook                     guestbook app (dynamic)
guestbook-dyn1    dynamic  applications  https://github.com/Rohitrajak1807/more_arlon  bundles/guestbook                     guestbook app (dynamic and updated)
guestbook-static  static   applications  (N/A)                                         (N/A)                                 guestbook app
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon cluster create --cluster-name test --repo-alias rrajak --repo-path basecluster1
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon basecluster validategit --repo-path basecluster1
Error: kustomization.yaml is missing
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon basecluster preparegit --repo-path basecluster1
preparation successful, cluster name: capi-quickstart
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon basecluster validategit --repo-path basecluster1
validation successful, cluster name: capi-quickstart
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon cluster create --cluster-name base1 --repo-path basecluster1
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon basecluster validategit --repo-path bcl --repo-alias rrajak
Error: kustomization.yaml is missing
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon basecluster preparegit --repo-path bcl --repo-alias rrajak
preparation successful, cluster name: capi-quickstart
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon cluster create --cluster-name base2 --repo-path bcl --repo-alias rrajak


➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon cluster create --cluster-name base3 --repo-path bcl --repo-alias 404-notfound
Error: given alias doesn't exist
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon profile list
NAME       GEN  TYPE     BUNDLES                            REPO-URL                                      REPO-PATH           OVRDS  TAGS        DESCRIPTION
dynamic-1  2    dynamic  [guestbook-static guestbook-dyn1]  https://github.com/Rohitrajak1807/arln-oss    profiles/dynamic-1  0      [examples]  dynamic test 1
static-1   2    static   [guestbook-static]                 (N/A)                                         (N/A)               0      [examples]  dynamic test 1
dynamic-2  2    dynamic  [guestbook-static guestbook-dyn1]  https://github.com/Rohitrajak1807/more_arlon  profiles/dynamic-2  0      [examples]  dynamic test 1
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon clusterspec create capi-eks-1 --api capi --cloud aws --type eks --kubeversion v1.22.9 --nodecount 2 --nodetype t2.medium --tags staging --desc "2 node eks for general purpose" --sshkey bogus --region us-west-1
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon cluster deploy --cluster-name first-cluster --profile static-1 --cluster-spec capi-eks-1
➜  arlon git:(private/Rohitrajak1807/git-ungit) bin/arlon cluster deploy --cluster-name not-first-cluster --profile static-1 --cluster-spec capi-eks-1 --repo-alias rrajak
```

### CODE REVIEW

**Comment by bcle** [comment]
Can you add a unit test, thanks.

**Comment by bcle** [comment]
The config file really shouldn't be stored in argocd config directory, since it's not managed by argocd code. We probably want an arlon specific directory.

**Comment by Rohitrajak1807** [comment]
> The config file really shouldn't be stored in argocd config directory, since it's not managed by argocd code. We probably want an arlon specific directory.

Any particular suggestions on where would the config dir be.

File: pkg/gitrepo/gitrepo.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+)
+
+var (
+	ErrNotFound     = errors.New("given alias doesn't exist")
+	ErrLoadCfgFile  = errors.New("failed to load config file")
+	ErrOverwriteCfg = errors.New("cannot overwrite config file")
+)
+
+func GetRepoCfgPath() (string, error) {
+	cfgDir, err := localconfig.DefaultConfigDir()
```
Comment (bcle) [inline pkg/gitrepo/gitrepo.go:35]: I mentioned this in another comment. I think arlon specific files should reside in arlon specific config directory.

**Review by bcle — CHANGES_REQUESTED** [review]

File: pkg/gitrepo/gitrepo.go (hunk 1 above)
Comment (bcle) [reply pkg/gitrepo/gitrepo.go:35]: > Any particular suggestions on where would the config dir be?
In the same root dir as the argocd config directory.

Comment (Rohitrajak1807) [reply pkg/gitrepo/gitrepo.go]: done

File: cmd/basecluster/preparegit.go [hunk 2]
```diff
@@ -13,12 +15,23 @@ func prepareGitBaseClusterCommand() *cobra.Command {
...
+	var repoAlias string
 	var repoRevision string
 	command := &cobra.Command{
 		Use:   "preparegit --repo-url repoUrl [--repo-revision revision] [--repo-path path]",
 		Short: "prepare base cluster directory in git",
 		Long:  "prepare base cluster directory in git",
 		RunE: func(c *cobra.Command, args []string) error {
+			if repoUrl == "" {
+				repoCtx, err := gitrepo.GetAlias(repoAlias)
+				if err != nil {
```
[the remark this answers is not in the document]
Comment (Rohitrajak1807) [reply cmd/basecluster/preparegit.go:27]: done

**Comment by Rohitrajak1807** [comment]
> Can you add a unit test, thanks.

done

File: cmd/gitrepo/unregister.go [hunk 3]
```diff
@@ -19,20 +19,8 @@ func unregister() *cobra.Command {
...
-			if err != nil {
-				return fmt.Errorf("%v: %w", gitrepo.ErrLoadCfgFile, err)
-			}
-			defer func(f *os.File) {
-				err := f.Close()
-				if err != nil {
-					fmt.Printf("failed to close config file, error: %v\n", err)
-				}
-			}(file)
+			file, err := gitrepo.ReadDefaultConfig()
```
Comment (bcle) [inline cmd/gitrepo/unregister.go:22]: Need to check the err value first

**Review by bcle — CHANGES_REQUESTED** [review]
One minor bug. Looks good otherwise.

File: cmd/gitrepo/unregister.go (hunk 3 above)
Comment (Rohitrajak1807) [reply cmd/gitrepo/unregister.go:22]: It is being checked right?
```go
repoAlias = args[0]
			file, err := gitrepo.ReadDefaultConfig()
			if err != nil {
				return err
			}
			defer argocdio.Close(file)
			repoCtxCfg, err := gitrepo.LoadRepoCfg(file)
			if err != nil {
				return fmt.Errorf("%v: %w", gitrepo.ErrLoadCfgFile, err)
			}
			if len(repoCtxCfg.Repos) == 0 {
				fmt.Println("no repositories registered")
				return
			}
```

Comment (Rohitrajak1807) [reply cmd/gitrepo/unregister.go:22]: We read the config and check the error, `argocdio.Close()` doesn't return an error, after loading the config too we check the error.

Comment (bcle) [reply cmd/gitrepo/unregister.go:22]: It's weird, I was seeing a different version of the code. Perhaps a caching problem. Looks ok now after a refresh.

**Review by bcle — APPROVED** [review]

**Comment by Rohitrajak1807** [comment]
merging...

**Pull request merged**